File: frontend/src/app/services/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})
export class AdminService {

  private api = 'http://localhost:8080/admin';

  constructor(private http: HttpClient) {}

  // deletePost(postId: number) {
  //   return this.http.delete(`${this.api}/delete-post/${postId}`);
  // }
  
  // deletePost(postId: number) {
  //   return this.http.delete(`http://localhost:8080/admin/delete-post/${postId}`);
  // }

  deletePost(postId: number) {
  return this.http.delete(`http://localhost:8080/admin/posts/${postId}`);
  }

  banUser(userId: number) {
    return this.http.post(`${this.api}/ban-user/${userId}`, {});
  }

  unbanUser(userId: number) {
  return this.http.post(`http://localhost:8080/admin/unban-user/${userId}`, {});
  }
  
  getAllUsers() {
  return this.http.get<any[]>('http://localhost:8080/admin/users');
  }

  getAllPosts() {
  return this.http.get<any[]>('http://localhost:8080/admin/posts');
  }

  hidePost(postId: number) {
  return this.http.post(`http://localhost:8080/admin/hide-post/${postId}`, {});
  }

  unhidePost(postId: number) {
  return this.http.post(`http://localhost:8080/admin/unhide-post/${postId}`, {});
  }
  
  getPostById(id: number) {
  return this.http.get(`http://localhost:8080/admin/posts/${id}`);
  }
  
  getUsersPaginated(lastId: number = 0, limit: number = 10) {
  return this.http.get<any[]>(`${this.api}/users/paginated?lastId=${lastId}&limit=${limit}`);
  }

  getPostsPaginated(lastId: number = 0, limit: number = 10) {
  return this.http.get<any[]>(`${this.api}/posts/paginated?lastId=${lastId}&limit=${limit}`);
  }
  
  
}

File: backend/src/main/java/com/dev/backend/services/AdminService.java
package com.dev.backend.services;

import com.dev.backend.entities.User;
import com.dev.backend.repositories.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.dev.backend.dto.AdminUserDTO;
import com.dev.backend.dto.AdminPostDTO;
import com.dev.backend.entities.Post;
import com.dev.backend.repositories.PostRepository;
import java.util.List;
import java.util.stream.Collectors;
import java.util.Optional;
import jakarta.transaction.Transactional;
import java.util.Comparator;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.PageRequest;



@Service
public class AdminService {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PostRepository postRepository;

    public List<AdminUserDTO> getAllUsers() {
    return userRepository.findAll()
        .stream()
        .filter(user -> !user.getEmail().equals("admin@gmail.com"))
        .map(user -> new AdminUserDTO(
            user.getId(),
            user.getFirstname(),
            user.getLastname(),
            user.getEmail(),
            user.isBanned()
        ))
        .toList();
    }
    
    public List<AdminPostDTO> getAllPosts() {
    return postRepository.findAll()
        .stream()
        // .filter(post -> !post.isHidden())
        .map(post -> new AdminPostDTO(
            post.getId(),
            post.getTitle(),
            post.getAuthor().getFirstname() + " " + post.getAuthor().getLastname(),
            post.isHidden()
        ))
        .toList();
    }

    public void hidePost(Long postId) {
    Post post = postRepository.findById(postId)
        .orElseThrow(() -> new RuntimeException("Post not found"));
    post.setHidden(true);
    postRepository.save(post);
    }

    public void unhidePost(Long postId) {
    Post post = postRepository.findById(postId)
        .orElseThrow(() -> new RuntimeException("Post not found"));
    post.setHidden(false);
    postRepository.save(post);
    }

    public void banUser(Long userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found"));
        user.setBanned(true);
        userRepository.save(user);
    }

    public void unbanUser(Long userId) {
    User user = userRepository.findById(userId)
        .orElseThrow(() -> new RuntimeException("User not found"));

    user.setBanned(false);
    userRepository.save(user);
}
    
    public void deletePostAsAdmin(Long postId) {
        Post post = postRepository.findById(postId)
            .orElseThrow(() -> new RuntimeException("Post not found"));
        postRepository.delete(post);
    }

    public Post getPostById(Long id) {
    return postRepository.findById(id)
        .orElseThrow(() -> new RuntimeException("Post not found"));
    }

    public List<AdminUserDTO> getUsersAfterId(Long lastId, int limit) {

    Pageable pageable = PageRequest.of(0, limit);

    return userRepository
        .findUsersAfterId(lastId == 0 ? null : lastId, pageable)
        .stream()
        .map(user -> new AdminUserDTO(
            user.getId(),
            user.getFirstname(),
            user.getLastname(),
            user.getEmail(),
            user.isBanned()
        ))
        .toList();
}

    public List<AdminPostDTO> getPostsAfterId(Long lastId, int limit) {

    Pageable pageable = PageRequest.of(0, limit);

    return postRepository
        .findPostsAfterId(lastId == 0 ? null : lastId, pageable)
        .stream()
        .map(post -> new AdminPostDTO(
            post.getId(),
            post.getTitle(),
            post.getAuthor().getFirstname() + " " + post.getAuthor().getLastname(),
            post.isHidden()
        ))
        .toList();
}



}

